rfd/000?-database-access.md

```diff
@@ -0,0 +1,512 @@
+
+2. Configure IAM policy in order to allow the user Teleport database service
+will use to connect to the database instance with the auth token. An example
+policy is shown below.
+
+```json
+{
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Effect": "Allow",
+            "Action": "rds-db:connect",
+            "Resource": [
+                "arn:aws:rds-db:us-west-1:1234567890:dbuser:db-ABCDEFGHIJKLMNOP/developer",
+            ]
+        }
+    ]
+}
+```
+
+Note that the above rule allows IAM authentication to a single database user,
+it also supports wildcards.
+
+https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/UsingWithRDS.IAMDBAuth.IAMPolicy.html
+
+3. Grant `rds_iam` role to the database users. For example, for PostgreSQL:
+
+```sql
+CREATE USER developer; 
+GRANT rds_iam TO developer;
+```
+
+https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/UsingWithRDS.IAMDBAuth.DBAccounts.html
+
+## RBAC
+
+Teleport role resource gets 3 new fields that allow to control access to 
+database instances as well as individual databases and database users.
+
+```yaml
+kind: role
+version: v3
+metadata:
+  name: developer
+spec:
+  allow:
+    # Label selectors for database instances this role has access to.
+    db_labels:
+      environment: ["dev", "stage"]
+    # Database names (within a database instance) this role has access to.
+    db_names: ["main", "metrics", "postgres"]
+    # Database users this role can log in as.
+    db_users: ["alice", "bob"]
+```
+
+## CLI
+
+To connect to a database instance, a user first must login using regular
+`tsh login` command:
+
+```sh
+$ tsh login --proxy=localhost:3080 --user=admin
+```
+
+Once logged in, they can see all available database using the following
+command:
+
+```sh
+$ tsh db ls
+Name            Description     Endpoint                                Labels
+--------------- --------------- --------------------------------------- ---------
+postgres        PostgreSQL 13.0 localhost:5432                          env=dev
+postgres-rds    PostgreSQL 12.4 postgres-rds.xxx.rds.amazonaws.com:5432 env=stage
+postgres-aurora PostgreSQL 11.6 postgres-rds.yyy.rds.amazonaws.com:5432 env=prod
+```
+
+The list will only show databases the logged in user has access to (as per 
+the database labels). For example, developer role shown above won't be able 
+to see the Aurora database instance:
+
+```sh
+$ tsh login --proxy=localhost:3080 --user=dev
+$ tsh db ls
+Name            Description     Endpoint                                Labels
+--------------- --------------- --------------------------------------- ---------
+postgres        PostgreSQL 13.0 localhost:5432                          env=dev
+postgres-rds    PostgreSQL 12.4 postgres-rds.xxx.rds.amazonaws.com:5432 env=stage
+```
+
+To log into a specific database instance, a user executes `tsh db login`
+command. This will retrieve the certificate with encoded database information
+in it:
+
+```sh
+$ tsh db login postgres
```

Comment:
How can I pick a user to login if I support multiple users?

Reply:
If you mean "database user" - do you think it needs to be specified here? Currently, you can just use any database user after login, as long it is allowed by RBAC rules, for example:

```sh
$ tsh db login postgres && eval $(tsh db env)
$ psql -U alice mydb     // allowed by role so will connect
$ psql -U bob mydb       // allowed by role so will connect
$ psql -U charlie mydb   // not allowed by role so will deny access
```

We could add a `--user` flag here and encode it into the certificate as well, but that would require the user to re-login if they want to use a different database user?

Reply:
`--user` is a bit overloaded so it'd have to be `--db-user` I think. This flow might be a bit cumbersome if they keep having to change users, though.

Reply:
@r0mant no need for a special flag then, add the explainer below to the rfd